pkg/sql/colexec/auto_incr.go

```diff
@@ -0,0 +1,307 @@
+
+	return nil
+}
+
+func UpdateInsertValueBatch(e engine.Engine, ctx context.Context, proc *process.Process, p *plan.InsertValues, bat *batch.Batch) error {
+	namePre := p.DbName + "_" + p.TblName + "_"
+	ColDefs := p.ExplicitCols
+	OrderColDefs(p.OrderAttrs, ColDefs)
+	return UpdateInsertBatch(e, ctx, proc, ColDefs, bat, namePre)
+}
+
+func GetRangeFromAutoIncrTable(e engine.Engine, ctx context.Context, proc *process.Process, bat *batch.Batch, ColDefs []*plan.ColDef, namePre string) ([]int64, error) {
+	taeEngine, ok := e.(moengine.TxnEngine)
+	if !ok {
+		return nil, errors.New("", "the engine is not tae")
+	}
+
+	dbSource, err := e.Database(ctx, "mo_catalog", proc.TxnOperator)
+	if err != nil {
+		return nil, err
+	}
+	rel, err := dbSource.Relation(ctx, "mo_increment_columns")
+	if err != nil {
+		return nil, err
+	}
+
+	var incrNum []int64
+
+	txnCtx, err := taeEngine.StartTxn(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, col := range ColDefs {
+		if !col.AutoIncrement {
+			continue
+		}
+		var d int64
+		if d, err = GetOneColRangeFromAutoIncrTable(rel, ctx, proc, bat, namePre+col.Name, i); err != nil {
+			if err2 := txnCtx.Rollback(); err2 != nil {
+				return nil, err2
+			}
+			return nil, err
+		}
+		incrNum = append(incrNum, d)
+	}
+	if err = txnCtx.Commit(); err != nil {
+		return nil, err
+	}
+	return incrNum, nil
+}
+
+func GetOneColRangeFromAutoIncrTable(rel engine.Relation, ctx context.Context, proc *process.Process, bat *batch.Batch, name string, pos int) (int64, error) {
+	var oriNum, maxNum, nullNum, cnt int64
+	oriNum = GetCurrentIndex(rel, ctx, proc, name)
+	if oriNum < 0 {
+		return 0, errors.New("", "GetIndex from auto_increment table fail")
+	}
+
+	vec := bat.Vecs[pos]
+	nullNum = int64(nulls.Length(vec.Nsp))
+	switch vec.Typ.Oid {
+	case types.T_int32:
+		vs := vec.Col.([]int32)
+		for rowIndex := 0; rowIndex < bat.Length(); rowIndex++ {
+			if vs[rowIndex] > int32(maxNum) {
+				maxNum = int64(vs[rowIndex])
+				cnt = 0
+			}
+			if nulls.Contains(vec.Nsp, uint64(rowIndex)) {
+				cnt++
+			}
+		}
+	case types.T_int64:
+		vs := vec.Col.([]int64)
+		for rowIndex := 0; rowIndex < bat.Length(); rowIndex++ {
+			if vs[rowIndex] > maxNum {
+				maxNum = vs[rowIndex]
+				cnt = 0
+			}
+			if nulls.Contains(vec.Nsp, uint64(rowIndex)) {
+				cnt++
+			}
+		}
+	}
+
+	maxNum += cnt
+	if oriNum+nullNum > maxNum {
+		maxNum = oriNum + nullNum
+	}
+	if err := UpdateAutoIncrTable(rel, ctx, maxNum, name); err != nil {
+		return 0, err
+	}
+	return oriNum, nil
+}
+
+func UpdateBatchImpl(ColDefs []*plan.ColDef, bat *batch.Batch, incrNum []int64) error {
+	pos := 0
+	for i, col := range ColDefs {
+		if !col.AutoIncrement {
+			continue
+		}
+		vec := bat.Vecs[i]
+		curNum := incrNum[pos]
+		pos++
+		switch vec.Typ.Oid {
+		case types.T_int32:
+			vs := vec.Col.([]int32)
+			for rowIndex := 0; rowIndex < bat.Length(); rowIndex++ {
+				if nulls.Contains(vec.Nsp, uint64(rowIndex)) {
+					nulls.Del(vec.Nsp, uint64(rowIndex))
```

Comment:
What is the logic here?   I don't quite understand ...

Reply:
when the data in auto_incr col is null, I need to rewrite the null value to a not null value